aggregator/src/aggregation/circuit.rs

```diff
@@ -290,12 +309,124 @@ impl<const N_SNARKS: usize> Circuit<Fr> for BatchCircuit<N_SNARKS> {
+                    }
+                    let transcript_init_state_halo2 = config
+                        .ecc_chip()
+                        .field_chip()
+                        .range()
+                        .gate()
+                        .assign_constant(&mut ctx, self.halo2_protocol.init_state)
+                        .expect("IntegerInstructions::assign_constant infallible");
+                    log::debug!("load transcript OK");
+                    let transcript_init_state_sp1 = config
+                        .ecc_chip()
+                        .field_chip()
+                        .range()
+                        .gate()
+                        .assign_constant(&mut ctx, self.sp1_protocol.init_state)
+                        .expect("IntegerInstructions::assign_constant infallible");
+                    log::info!("populating constants OK");
+
+                    // Commitments to the preprocessed polynomials.
+                    for preprocessed_polys in preprocessed_poly_sets.iter() {
+                        let mut preprocessed_check_1 =
+                            config.flex_gate().load_constant(&mut ctx, Fr::ONE);
+                        let mut preprocessed_check_2 =
+                            config.flex_gate().load_constant(&mut ctx, Fr::ONE);
+                        for ((commitment, comm_halo2), comm_sp1) in preprocessed_polys
+                            .iter()
+                            .zip_eq(preprocessed_polys_halo2.iter())
+                            .zip_eq(preprocessed_polys_sp1.iter())
+                        {
+                            let check_1 =
+                                config
+                                    .ecc_chip()
+                                    .is_equal(&mut ctx, &commitment, comm_halo2);
+                            let check_2 =
+                                config.ecc_chip().is_equal(&mut ctx, &commitment, comm_sp1);
+                            preprocessed_check_1 = config.flex_gate().and(
+                                &mut ctx,
+                                Existing(preprocessed_check_1),
+                                Existing(check_1),
+                            );
+                            preprocessed_check_2 = config.flex_gate().and(
+                                &mut ctx,
+                                Existing(preprocessed_check_2),
+                                Existing(check_2),
+                            );
+                        }
+                        let preprocessed_check = config.flex_gate().or(
+                            &mut ctx,
+                            Existing(preprocessed_check_1),
+                            Existing(preprocessed_check_2),
+                        );
+                        config
+                            .flex_gate()
+                            .assert_is_const(&mut ctx, &preprocessed_check, Fr::ONE);
```

Comment:
ok i see.  Here the design is, you verify sp1 proof & protocol, you verify halo2 proof & protocol, and you assert one of them to be true.

this design is ok.  But not efficient.  We should optimize it later.   (Actually using this design do you really need `loaded_preprocessed_as_witness`?) 

the optimized design is,  we could make sp1 chunk protocol and halo2 chunk protocol, the circuits shapes be same (only fixed_columns/vk be different).  So the `Plonk::<PCS>::succinct_verify` is called once, not twice.   And assert     protocol_digest outside.

Reply:
oh not. let me read

Reply:
oh, I thought chunks in a batch could be _either_ sp1 _or_ halo2. Isn't that true?

Reply:
Do you mean, a batch is _all_ sp1 _or all_ halo2?